### PR TITLE
Fix routing loop prevention, MagicDNS forwarding over Tailscale.

### PR DESCRIPTION
**Pull request by bradfitz**
Fixes tailscale/tailscale#2102
Updates tailscale/tailscale#1809

### CODE REVIEW

File: cmd/tailscale/main.go [hunk 1]
```diff
@@ -328,6 +329,17 @@ func (a *App) runBackend() error {
 				if service != 0 {
 					jni.DeleteGlobalRef(env, service)
 				}
+				netns.SetAndroidProtectFunc(func(fd int) error {
+					return jni.Do(a.jvm, func(env *jni.Env) error {
+						cls := jni.GetObjectClass(env, s)
+						m := jni.GetMethodID(env, cls, "protect", "(I)Z")
+						ret, err := jni.CallBooleanMethod(env, s, m, jni.Value(fd))
+						if err != nil || ret == false {
+							log.Printf("[unexpected] VpnService.protect(%d) = %v, %v", fd, ret, err)
```
Comment (crawshaw) [inline cmd/tailscale/main.go:338]: Leave a comment explaining why you don't return this error?

**Review by crawshaw — APPROVED** [review]

File: cmd/tailscale/main.go (hunk 1 above)
Comment (bradfitz) [reply cmd/tailscale/main.go]: PTAL

**Review by crawshaw — APPROVED** [review]

**Review by danderson — APPROVED** [review]

**Comment by bradfitz** [comment]
@eliasnaur, happy to address any feedback you have if I screwed something up with JNI. Or feel free to fix it yourself too.

**Comment by eliasnaur** [comment]
LGTM, thanks!

**Pull request merged**